### PR TITLE
[CLOB-1020] [CLOB-1048] Logging Library with Contextual Tags

### PR DESCRIPTION
**Pull request by jonfung-dydx**
1. Introduce a new logging entrypoint functions in `protocol/lib/log/utils.go`. 
- These functions take in the context `ctx` object and grab the logger from the context, which carries a set of k:v tag values for logs.
- New method `AddPersistentTagsToLogger` returns a new context `ctx` object with specified k:v tags attached onto the logger.
- All subsequent logging calls using the specified ctx object will contain previous tags added with `AddPersistentTagsToLogger`
2. `protocol/lib/log/constants.go` constants for metric tag keys
3. Migrate callsites in [protocol/x/clob/keeper/](https://github.com/dydxprotocol/v4-chain/pull/872/files#diff-ae3a9d0d6029ff89d66f4bf9a308ce4d067f532bf0b5ee25d9bb0ebc9a5f197f) to use new library

### CODE REVIEW

File: .github/workflows/protocol-build-and-push-snapshot.yml
Comment (jonfung-dydx) [inline .github/workflows/protocol-build-and-push-snapshot.yml:6]: will remove these before merge

File: protocol/x/clob/abci.go
Comment (jonfung-dydx) [inline protocol/x/clob/abci.go:133]: all abci.go functions have tags on them

File: protocol/x/clob/ante/clob.go
Comment (jonfung-dydx) [inline protocol/x/clob/ante/clob.go:67]: Note we add persistent tags for checktx here

File: protocol/x/clob/keeper/msg_server_cancel_orders.go
Comment (jonfung-dydx) [inline protocol/x/clob/keeper/msg_server_cancel_orders.go:28]: delivertx persistent tags

File: protocol/lib/log/constants.go
Comment (lucas-dydx) [inline protocol/lib/log/constants.go:35]: why do these tag values use underscores (or backslashes)?

why are module names like `x/clob` defined in same const def as callbacks?

File: protocol/lib/log/utils.go
Comment (lucas-dydx) [inline protocol/lib/log/utils.go:11]: is it worth validating that `keyvals` is an even length?

What happens if it's an odd-length? Should we note this in the comment?

File: protocol/lib/log/utils.go
Comment (lucas-dydx) [inline protocol/lib/log/utils.go:27]: Assuming u meant this

```suggestion
// Please try to define a new message and use `ErrorLogWithError` instead.
```

File: protocol/x/clob/abci.go
Comment (lucas-dydx) [inline protocol/x/clob/abci.go:28]: nit: each arg on it's own line

File: protocol/x/clob/abci.go
Comment (lucas-dydx) [inline protocol/x/clob/abci.go:133]: Shouldn't this be block height + 1 since it's `PrepareCheckState` for the next block?

File: protocol/x/clob/ante/clob.go
Comment (lucas-dydx) [inline protocol/x/clob/ante/clob.go:70]: isn't this also block height + 1 since it's not deliver tx?

File: protocol/x/clob/keeper/msg_server_cancel_orders.go
Comment (lucas-dydx) [inline protocol/x/clob/keeper/msg_server_cancel_orders.go:34]: is `msg` just a pointer in the resulting log? or do the fields get logged properly

**Review by lucas-dydx — APPROVED** [review]

File: protocol/lib/log/constants.go
Comment (jonfung-dydx) [reply protocol/lib/log/constants.go:35]: To retain backwards compatibility with the existing tags. Existing tags are in underscore delimited except for the module names, which are `x/clob` and `x/vesting`.

File: protocol/lib/log/utils.go
Comment (jonfung-dydx) [reply protocol/lib/log/utils.go:11]: I think we should adopt the same mentality as the underlying library, which doesn't perform validation. In an odd number case, the underlying library + this function discards the last value. I think Logging/metrics libraries should not be able to affect application behavior, so in this case I think we shouldn't panic.

File: protocol/x/clob/abci.go
Comment (jonfung-dydx) [reply protocol/x/clob/abci.go:28]: i think it's a bit more clear here since it's `key, val` alternating here

File: protocol/x/clob/keeper/msg_server_cancel_orders.go
Comment (jonfung-dydx) [reply protocol/x/clob/keeper/msg_server_cancel_orders.go:34]: Looks fine to me
Screenshot 2024-01-04 at 12 35 44 AM

File: protocol/lib/log/constants.go
Comment (jakob-dydx) [inline protocol/lib/log/constants.go:45]: nit, but isn't this PascalCase, not camelCase ;)

File: protocol/lib/log/utils.go
Comment (jakob-dydx) [inline protocol/lib/log/utils.go:41]: Not super opinionated about this, but have you considered alternative names for this function?

Was thinking something like `PutTagsToLogger` could be helpful, since if this function is called twice it is possible to overwrite previous tag values. The idea of a put operation is common and could make the internals of this extra clear. But yeah, your call

**Review by jakob-dydx — APPROVED** [review]

**Pull request merged**
